clap_derive/src/derives/args.rs

```diff
@@ -51,10 +52,188 @@ pub(crate) fn derive_args(input: &DeriveInput) -> Result<TokenStream, syn::Error
+                .iter()
+                .map(|variant| {
+                    let item =
+                        Item::from_args_enum_variant(variant, item.casing(), item.env_casing())?;
+                    Ok((item, variant))
+                })
+                .collect::<Result<Vec<_>, syn::Error>>()?;
+
+            gen_for_enum(&item, ident, &input.generics, &variant_items)
+        }
         _ => abort_call_site!("`#[derive(Args)]` only supports non-tuple structs"),
     }
 }
 
+pub(crate) fn gen_for_enum(
+    _item: &Item,
+    item_name: &Ident,
+    generics: &Generics,
+    variants: &[(Item, &Variant)],
+) -> Result<TokenStream, syn::Error> {
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let app_var = Ident::new("__clap_app", Span::call_site());
+    let mut augmentations = TokenStream::default();
+    let mut augmentations_update = TokenStream::default();
+
+    let mut constructors = TokenStream::default();
+    let mut updaters = TokenStream::default();
+
+    for (item, variant) in variants.iter() {
+        let Fields::Named(ref fields) = variant.fields else {
+            abort! { variant.span(),
+                "`#[derive(Args)]` only supports named enum variants if used on an enum",
+            }
+        };
+        let group_id = item.group_id();
+
+        let conflicts = variants
+            .iter()
+            .filter_map(|(_, v)| {
+                if v.ident == variant.ident {
+                    None
+                } else {
+                    Some(Name::Derived(v.ident.clone()))
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let fields = collect_args_fields(item, fields)?;
+
+        let augmentation = gen_augment(&fields, &app_var, item, &conflicts, false)?;
+        let augmentation = quote! {
+            let #app_var = #augmentation;
+        };
+        let augmentation_update = gen_augment(&fields, &app_var, item, &conflicts, true)?;
+        let augmentation_update = quote! {
+            let #app_var = #augmentation_update;
+        };
+
+        augmentations.extend(augmentation);
+        augmentations_update.extend(augmentation_update);
+
+        let variant_name = &variant.ident;
+        let genned_constructor = gen_constructor(&fields)?;
+        let constructor = quote! {
+            if __clap_arg_matches.contains_id(#group_id) {
+                let v = #item_name::#variant_name #genned_constructor;
+                return ::std::result::Result::Ok(v)
+            }
+        };
+
+        constructors.extend(constructor);
+
+        let genned_updater = gen_updater(&fields, false)?;
+
+        let field_names = fields
+            .iter()
+            .map(|(field, _)| field.ident.as_ref().unwrap());
+        let updater = quote! {
+
+            if __clap_arg_matches.contains_id(#group_id) {
+                let #item_name::#variant_name { #( #field_names ),* } = self else {
+                    unreachable!();
```

Comment:
i guess this might fail if the current variant is different, should it be possible to "change" variants using `update_from_arg_matches_mut`?

Reply:
This should probably be an error, or instead of trying to update we'd attempt to parse all fields of the "other" variant and replace self with that.

Reply:
for a slightly more concrete discussion:
this currently generates:

```rust
fn update_from_arg_matches_mut(
    &mut self,
    __clap_arg_matches: &mut clap::ArgMatches,
) -> ::std::result::Result<(), clap::Error> {
    #![allow(deprecated)]
    if __clap_arg_matches.contains_id("A") {
        let Source::A { a, aaa } = self else {
            unreachable!();
        };
        if __clap_arg_matches.contains_id("a") {
            *a = __clap_arg_matches.remove_one::<bool>("a").ok_or_else(|| {
                clap::Error::raw(
                    clap::error::ErrorKind::MissingRequiredArgument,
                    concat!("The following required argument was not provided: ", "a"),
                )
            })?
        }
        if __clap_arg_matches.contains_id("aaa") {
            *aaa = __clap_arg_matches
                .remove_one::<bool>("aaa")
                .ok_or_else(|| {
                    clap::Error::raw(
                        clap::error::ErrorKind::MissingRequiredArgument,
                        concat!("The following required argument was not provided: ", "aaa"),
                    )
                })?
        };
    }
    if __clap_arg_matches.contains_id("B") {
        let Source::B { b } = self else {
            unreachable!();
        };
        if __clap_arg_matches.contains_id("b") {
            *b = __clap_arg_matches.remove_one::<bool>("b").ok_or_else(|| {
                clap::Error::raw(
                    clap::error::ErrorKind::MissingRequiredArgument,
                    concat!("The following required argument was not provided: ", "b"),
                )
            })?
        };
    }
    ::std::result::Result::Ok(())
}
```

for my test enum:

```rust
#[derive(clap::Args, Clone, Debug, PartialEq, Eq)]
enum Source {
    A {
        #[arg(short)]
        a: bool,
        #[arg(long)]
        aaa: bool,
    },
    B {
        #[arg(short)]
        b: bool,
    },
}
```

Reply:
I would look to subcommands for a starting point in designing how to model this.

I also am really tempted to remove all of the `update` code...